### PR TITLE
Shutter fork v2

### PR DESCRIPTION
**Pull request by gswirski**

### CODE REVIEW

**Comment by simonmitchell** [comment]
@gswirski will try and leave some comments on this tomorrow ☺️ Should it still have the do not merge label?

**Comment by gswirski** [comment]
@simonmitchell sorry, it's not meant to go into the main Rocc repo. It's out of date, and not limited to the "closeSession" issue we've talked over email.

I use this PR to highlight my changes to Rocc. I still need to figure out which changes are helpful, and which are just my misunderstanding of Rocc internals. Some things are already out of date (e.g. you've already removed ThunderRequest sub-framework). Feel free to comment on what you think would make sense in the upstream repo, and I can start working on cleaner patches.

At the high level, this PR contains the following changes:
- no ThunderRequest sub-module to make App Store Connect happy - already fixed in the main repo
- public initializers for Aperture and Exposure Compensation
- a way to highlight unsupported shooting modes. I chose .loop as a temporary hack, but we should figure out how to properly handle bracketing etc.
- separate callback for TCP socket disconnections. As discussed earlier, this can happen when the app is in the background for too long and is completely normal. I want to re-try the connection before deciding that the camera has disconnected.
- simpler device connectivity notifier. In my recent experience, Reachability contributed a bit too much noise, and Wi-Fi changes were enough
- minor live view optimization not to send obsolete frames to the delegate
- "closeSession" command when the camera indicates another session is open - needs to be tested on a7R IV
- 3 retries when the handshake fails - not sure if the best way to handle failures

File: Rocc/DeviceConnectivityNotifier.swift [hunk 1]
```diff
@@ -95,23 +105,25 @@ public final class DeviceConnectivityNotifier {
         isListening = true
         reachability = _reachability
         initialSSID = Reachability.currentWiFiSSID
-        
+
+        os_log("Initial SSID %s", log: self.logger, type: .debug, String(describing: initialSSID))
+
         reachability?.start(callback: { [weak self] (flags) in
-            self?.handle(flags: flags)
+            //self?.handle(flags: flags)
```
Comment (simonmitchell) [inline Rocc/DeviceConnectivityNotifier.swift:112]: I'm a bit cautious to remove this :D Can we move the Reachability changes into their own PR and discuss there when you PR this back into `simonmitchell/rocc`? :)

File: Rocc/LiveViewStreaming.swift [hunk 2]
```diff
@@ -420,17 +420,28 @@ public final class LiveViewStream: NSObject {
         } else if !isPacketedStream {
             isPacketedStream = true
         }
-        
+
+        var lastImage: Image?
```
Comment (simonmitchell) [inline Rocc/LiveViewStreaming.swift:424]: Again, when PRing into `simonmitchell/rocc` would be good to get this separated off :D Also wondering if it would be better to reverse this array and use `first(where:)` rather than looping all the way through to find the last element? I'm guessing it won't make much difference as we can expect maybe max 2 payloads at a time?

File: Rocc/Manufacturer Implementations/Sony/PTP IP Camera/SonyPTPIPCamera.swift [hunk 3]
```diff
@@ -443,24 +460,66 @@ extension SonyPTPIPDevice: Camera {
...
             self?.handlePTPIPEvent(event)
         }
         ptpIPClient?.onDisconnect = { [weak self] in
             self?.onDisconnected?()
         }
     }
-    
+
+    typealias WorkSuccess = (_ retriable: Bool) -> Bool
+    typealias WorkBlock = (_ attemptRetry: @escaping WorkSuccess, _ attemptNumber: Int) -> Void
```
Comment (simonmitchell) [inline Rocc/Manufacturer Implementations/Sony/PTP IP Camera/SonyPTPIPCamera.swift:508]: This is really nice! I think it has scope to be added under `/helpers` as a free function like `max` e.t.c. It doesn't obviously act on anything so I think it is valid to have it as a free (global) function 😄

File: Rocc/Manufacturer Implementations/Sony/PTP IP Camera/SonyPTPIPCamera.swift [hunk 4]
```diff
@@ -205,10 +212,19 @@ internal final class SonyPTPIPDevice: SonyCamera {
         
         performSdioConnect(completion: { [weak self] (error) in
             guard let self = self else { return }
+
+            if let ptpError = error as? PTPError, case PTPError.commandRequestFailed(.nikon_batteryLow) = ptpError {
```
Comment (simonmitchell) [inline Rocc/Manufacturer Implementations/Sony/PTP IP Camera/SonyPTPIPCamera.swift:216]: Like you were saying via email, let's rename this `.nikon_batteryLow` for now ☺️ At some point I'm going to probably try name-spacing the return codes somehow to manufacturers, but we don't need to worry about that for now!

**Comment by simonmitchell** [comment]
Hey man, thanks for sharing this with me! I'd like to definitely get the PTP/IP connection stuff into ROCC if your kind beta-tester says it's resolved. The reachability stuff I'm less sure about, but if we can maybe look to making it a configuration property on `DeviceConnectivityNotifier`? 

Maybe add a `configuration` property to it's initialisation something along the lines of:

```
class DeviceConnectivityNotifier {
   struct Options: OptionSet {

       public let rawValue: Int
        
        public init(rawValue: RawValue) {
            self.rawValue = rawValue
        }

        public static let listenForReachabilityChanges = Options(rawValue: 1 << 0)
   }
}
```

And then default it to include that option, so you can opt out but if others (And I) want to use `Reachability` we still can? :) 

I'm happy for the rest of it to go in, that's the only one I have some qualms about! Again, awesome work thank you so much!
